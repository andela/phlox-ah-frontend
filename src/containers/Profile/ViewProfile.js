/* eslint-disable no-script-url */
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import Avatar from 'react-avatar';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { viewProfile } from '../../requests/ProfileRequest';
import FollowList from '../../components/FollowList/FollowList';
import Tags from '../../components/Tags/Tags';
import { getFollowings, getFollowers } from '../../requests/FollowRequests';
import AuthorsArticle from '../AuthorsArticle/AuthorsArticle';

import './ViewProfile.scss';
import './Profile.scss';

/**
 * @class ViewProfile
 * @extends {Component}
 */
class ViewProfile extends Component {
  /**
   * @description - This method runs after component has been mounted
   * @returns {object} articles
   * @memberof ViewProfile
   */
  componentDidMount() {
    this.props.getFollowings();
    this.props.getFollowers();
  }

  /**
   *
   * @description - This method renders the jsx for this component
   * @returns {jsx} - jsx
   * @memberof ViewProfile
   */
  render() {
    const tags = ['Religion', 'Sports', 'Technology', 'Music', 'Art', 'Software', 'Finance'];
    const listOfFollowings = (
      this.props.followings.slice(0, 3).map((following, i) => <FollowList
      key={i}>{following.Profile && following.Profile.profileImage ? <img src={following.Profile.profileImage} alt="Profile Image" className="circle responsive-img follow-pic"/>
        : <Avatar name={following.username} size="50" round={true} />}{following.username}</FollowList>));
    const listOfFollowers = (
      this.props.followers.slice(0, 3).map((followers, i) => <FollowList
      key={i}>{followers.Profile && followers.Profile.profileImage ? <img src={followers.Profile.profileImage} alt="Profile Image" className="circle responsive-img follow-pic"/>
        : <Avatar name={followers.username} size="50" round={true} />}{followers.username}</FollowList>));
    const tagList = tags.map((tag, i) => <Tags key={i}>{tag}</Tags>);

    return (
      <div className="profile">
        <div className="profile-wrapper">
          <div className="profile-info tablet">
            <div className="info-wrapper">
              <div className="pi-photo">
                <div className="profile-image">
                  {
                    this.props.profile.profileImage ? <img src={this.props.profile.profileImage} alt="Profile Image" className="circle responsive-img"/>
                      : <Avatar name={this.props.user.username} size="75" round={true} />
                  }
                </div>
              </div>
              <div className="pi-detail">
                <div className="username">
                  {`${this.props.profile.lastName} ${this.props.profile.firstName}`}
                </div>
                <table className="profile-data">
                  <tbody>
                    <tr className="data-row">
                      <td>Email: </td>
                      <td>{this.props.user.email}</td>
                    </tr>
                    <tr className="data-row">
                      <td>Bio: </td>
                      <td>
                        {this.props.profile.bio}
                      </td>
                    </tr>
                    <tr className="data-row">
                      <td>Contact: </td>
                      <td>
                        {this.props.profile.contact}
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>

            <div className="followers">
              <div className="follow">
                <span className="header">
                  Authors I Follow
                </span>
                <ul className="authors">
                  {listOfFollowings}
                  {!this.props.followings.length && <p>You are not following any author</p>}
                </ul>
                <div className="more">
                  <Link to="/following">View more</Link>
                </div>
              </div>
              {/* end of follow */}
              <div className="following">
                <span className="header">
                  Authors Following Me
                </span>
                <ul className="authors">
                  {listOfFollowers}
                  {!this.props.followers.length && <p>You do not have any followers</p>}
                </ul>
                 <div className="more">
                  <Link to="/followers">View more</Link>
                </div>
              </div>
              {/* end of following */}
            </div>
            {/* end of followers */}

            <AuthorsArticle {...this.props} />

          </div>
          {/* end of profile info */}
          <div className="profile-edit">
            <div className="edit-profile-link">
              <Link to='/profile/edit'>Edit Profile</Link>
            </div>
            <div className="pref-heading">
              My Article Preferences
            </div>
            <div className="tags">
              <ul className="tags-wrapper">
                {tagList}
              </ul>
            </div>
          </div>
        </div>


      </div>
    );
  }
}

const mapStateToProps = state => ({
  profile: state.profile,
  user: state.user,
  followings: state.followUser.followings,
  followers: state.followUser.followers
});

ViewProfile.propTypes = {
  viewProfile: PropTypes.func,
  firstName: PropTypes.string,
  lastName: PropTypes.string,
  gender: PropTypes.string,
  bio: PropTypes.string,
  contact: PropTypes.string,
  profileImage: PropTypes.string,
  profile: PropTypes.object,
  user: PropTypes.object,
  getFollowings: PropTypes.func.isRequired,
  getFollowers: PropTypes.func.isRequired,
  followings: PropTypes.array,
  followers: PropTypes.array,
};

export default connect(mapStateToProps, {
  viewProfile, getFollowings, getFollowers
})(ViewProfile);
