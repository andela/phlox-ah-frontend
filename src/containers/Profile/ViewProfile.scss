
$primary-color: #06576D;


@import '../../styles/variables';
div.profile {
  width: 100%;
  margin: 30px 0;

  a {
    text-decoration: none;
  }
}

div.profile div.profile-wrapper {
  margin: auto;
  width: 90%;
  display: flex;
  flex-wrap: wrap;
}

div.profile-wrapper div.profile-info,
div.profile-wrapper div.profile-edit {
  padding: 10px;
  background: rgba(0,0,0,0.04);
  margin: 5px;
}

div.profile-wrapper div.profile-info {
  width: 70%;
  display: flex;
  flex-wrap: wrap;
  min-width: 320px;
}

div.profile-info div.info-wrapper {
  display: inherit;
}

div.profile-info div.pi-photo {
  width: 100px;
}

div.pi-photo span.photo,
div.pi-photo span.change-photo {
  display: block;
}

div.pi-photo span.photo {
  border-radius: 50%;
  background: #ccc;
  width: 80px;
  height: 80px;
  margin: auto;
}

div.pi-photo span.change-photo {
  text-align: center;
  padding: 5px 0;
  display: none;
}

div.profile-info div.pi-detail {
  flex: 1;
  min-width: 300px;
  margin-left: 50px;
}

div.pi-detail div.username {
  font-size: 30px;
  font-weight: 600;
}

table.profile-data tr.data-row {
  border-bottom: none;
}

table.profile-data tr.data-row td:first-child{
  font-weight: 600;
  vertical-align: top;
}

table.profile-data tr.data-row td:nth-child(2){
  color: #777;
}

table.profile-data tr.data-row td{
  padding: 2px 3px;
  font-size: 16px;
  letter-spacing: 0.6px;
  line-height: 1.5;
}

div.profile-info div.followers {
  width: 100%;
  display: flex;
  flex-wrap: wrap;
  margin-top: 30px;
  padding-top: 6px;
}

div.followers div.follow,
div.followers div.following {
  flex-basis: 50%;
  padding: 5px;
  min-width: 300px;

  span.header {
    font-size: 20px;
    font-weight: 600;
    color: $primary-color;
    display: block;
    margin-left: 11px;
  }

  ul.authors {
    margin: 4px;
  }

  ul.authors li.author {
    display: flex;
    padding: 4px 0;
  }

  li.author a {
    display: inherit;
    text-decoration: none;
  }

  li.author div.photo {
    width: 50px;
    height: 50px;
    background: #ccc;
    border-radius: 50%;
    margin: 2px;
    margin-right: 5px;
  }

  li.author div.name {
    font-size: 16px;
    color: #777;
    line-height: 3;
    text-align: left;
    flex: 1;
    margin-left: 10px;
  }

  div.more {
    text-align: center;
    font-weight: 600;

    a {
      color: #039be5;
    }
  }
}

div.followers div.following {
  border-left: 1px solid #ccc;
  padding-left: 15px;
}

div.profile-wrapper div.profile-edit {
  flex: 1;
  min-width: 220px;
  align-self: self-start;

  div.edit-profile-link {
    text-align: center;
    font-size: 20px;
    font-weight: 600;
    text-decoration: underline;
    letter-spacing: 0.5px;
    color: #039be5;
    a {
      color: $themeColor;
      &:hover {
        color: darken($themeColor, 20);
        transition: 0.5s;
      }
    }
  }

  div.pref-heading {
    font-size: 18px;
    text-align: center;
    font-weight: 600;
    margin-top: 4px;
    color: #666;
  }

  div.tags ul.tags-wrapper {
    display: flex;
    flex-wrap: wrap;
    margin: 15px 0;
    max-height: 200px;
    overflow-y: auto;
  }

  ul.tags-wrapper li.tag {
    background: $primary-color;
    color: #fff;
    padding: 4px 13px;
    border-radius: 3px;
    font-size: 14px;
    letter-spacing: 0.5px;
    margin: 4px;
  }

  li.tag span.remove-tag {
    display: inline-block;
    margin-left: 8px;
    cursor: pointer;
  }
}


@media screen and (max-width: 890px) {
 div.profile-wrapper div.profile-info.tablet {
    width: 100%;
  }
}


@media screen and (max-width: 700px) {
  div.followers div.follow,
  div.followers div.following {
    margin: auto;
  }
  div.followers div.following {
    border: none;
    margin-top: 20px
  }
}


