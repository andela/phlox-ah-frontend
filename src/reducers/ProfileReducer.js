import { NEW_PROFILE, VIEW_PROFILE } from '../actionTypes/ProfileConstants';
import { asyncActionName } from '../util/AsyncUtil';

const initialState = {
  firstName: '',
  lastName: '',
  gender: '',
  bio: '',
  contact: '',
  profileImage: ''
};

const profileReducer = (state = initialState, action) => {
  switch (action.type) {
    case asyncActionName(NEW_PROFILE).success:
      return { ...action.payload.profile };
    case asyncActionName(VIEW_PROFILE).success:
      return { ...state, ...action.payload.profile.profile };
    default:
      return state;
  }
};

export default profileReducer;
