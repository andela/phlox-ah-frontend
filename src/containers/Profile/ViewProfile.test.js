import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ViewProfile from './ViewProfile';
import ProfileReducer from '../../reducers/ProfileReducer';

const props = {
  profile: {
    firstName: '',
    lastName: '',
    gender: '',
    bio: '',
    contact: '',
    profileImage: 'https://via.placeholder.com/300?text=AuthorsHaven'
  },
  user: {
    username: '',
    email: ''
  },

};

const mockStore = configureMockStore([thunk]);
const store = mockStore({
  ...props,
  viewProfile: jest.fn(),
  followUser: [{ followings: [{}], followers: [{}] }]
});

let component;
let myComponent;


describe('<ViewProfile/>', () => {
  beforeEach(() => {
    component = shallow(
      <Provider store={store}>
        <ViewProfile />
      </Provider>
    );
    myComponent = component.dive({ context: { store } }).dive();
  });
  it('should render without throwing an error', () => {
    expect(component).toMatchSnapshot();
  });
  it('should have div.profile tag', () => {
    expect(myComponent.find('div.profile').exists()).toBe(true);
  });
  it('should have img tag', () => {
    expect(myComponent.find('img').exists()).toBe(true);
  });
  it('should have a div with class pi-photo', () => {
    expect(myComponent.find('div.pi-photo').exists()).toBe(true);
  });
  it('should have a FollowList component', () => {
    expect(myComponent.find('FollowList').exists()).toBe(true);
  });
  it('should have a Tags component', () => {
    expect(myComponent.find('Tags').exists()).toBe(true);
  });
  it('should have a Link component', () => {
    expect(myComponent.find('Link').exists()).toBe(true);
  });
  it('should have ul with class authors', () => {
    expect(myComponent.find('ul.authors').exists()).toBe(true);
  });
  it('should have ul with class tags-wrapper', () => {
    expect(myComponent.find('ul.tags-wrapper').exists()).toBe(true);
  });
  it('should have table.profile-data tag', () => {
    expect(myComponent.find('table.profile-data').exists()).toBe(true);
  });
  it('should have tr with class data-row', () => {
    expect(myComponent.find('tr.data-row').exists()).toBe(true);
  });

  it('should dispatch an action to view profile', () => {
    const payload = {
      profile: props,
      message: 'Profile fetched successfully',
      success: true
    };
    const action = {
      type: 'VIEW_PROFILE_SUCCESS',
      payload
    };
    const newState = ProfileReducer({}, action);
    expect(newState).toEqual({ ...props.profile });
  });
});
