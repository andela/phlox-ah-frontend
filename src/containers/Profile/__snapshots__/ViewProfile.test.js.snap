// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewProfile/> should render without throwing an error 1`] = `<Connect(ViewProfile) />`;
